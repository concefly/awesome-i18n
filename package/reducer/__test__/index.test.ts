import { LoadResult, LoadResultItem } from "ai18n-type";
import Reducer from "../lib";

describe("reduce", () => {
  test("合并相同 key", () => {
    const result = new Reducer().reduce(
      new LoadResult([
        new LoadResultItem("香蕉", "a"),
        new LoadResultItem("香蕉", "b"),
        new LoadResultItem("香蕉"),
        new LoadResultItem("苹果 {name}"),
        new LoadResultItem("西瓜"),
        new LoadResultItem("西瓜"),
      ]),
      new LoadResult([])
    );

    expect(result).toMatchSnapshot();
  });

  test("新增 & 删除", () => {
    const result = new Reducer().reduce(
      new LoadResult([
        new LoadResultItem("香蕉", "c"),
        new LoadResultItem("苹果 {name}"),
      ]),
      new LoadResult([
        new LoadResultItem("香蕉", "a"),
        new LoadResultItem("香蕉", "b"),
        new LoadResultItem("雪梨"),
      ])
    );

    expect(result).toMatchSnapshot();
  });

  test("冲突场景 - 保留原来的翻译", () => {
    const result = new Reducer().reduce(
      new LoadResult([new LoadResultItem("香蕉", undefined)]),
      new LoadResult([new LoadResultItem("香蕉", undefined, { msg: "banana" })])
    );

    expect(result).toMatchSnapshot();
  });

  test("冲突场景 - 冲突文案的新上下文复用翻译", () => {
    const result = new Reducer().reduce(
      new LoadResult([new LoadResultItem("香蕉", "a")]),
      new LoadResult([new LoadResultItem("香蕉", undefined, { msg: "banana" })])
    );
    expect(result).toMatchSnapshot();
  });
});

describe("extract", () => {
  it("extract", () => {
    const result = new Reducer().extract({
      香蕉: "香蕉 {name}",
      苹果: "{description, select, a{apple_a} b{apple_b}}",
    });

    expect(result).toMatchSnapshot();
  });
});
