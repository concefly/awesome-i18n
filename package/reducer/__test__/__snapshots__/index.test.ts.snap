// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract extract 1`] = `
ExtractResult {
  "list": Array [
    ExtractResultItem {
      "description": "other",
      "extra": Object {
        "msg": "香蕉 {name}",
      },
      "key": "香蕉",
    },
    ExtractResultItem {
      "description": "a",
      "extra": Object {
        "msg": "apple_a",
      },
      "key": "苹果",
    },
    ExtractResultItem {
      "description": "b",
      "extra": Object {
        "msg": "apple_b",
      },
      "key": "苹果",
    },
  ],
}
`;

exports[`reduce 冲突场景 - 保留原来的翻译 1`] = `
ReduceResult {
  "data": Map {
    "香蕉" => ReduceResultItem {
      "value": ICUItem {
        "dumpStr": "{description, select, other{banana}}",
        "translates": Array [],
      },
    },
  },
}
`;

exports[`reduce 冲突场景 - 冲突文案的新上下文复用翻译 1`] = `
ReduceResult {
  "data": Map {
    "香蕉" => ReduceResultItem {
      "value": ICUItem {
        "dumpStr": "{description, select, a{banana} other{banana}}",
        "translates": Array [],
      },
    },
  },
}
`;

exports[`reduce 合并相同 key 1`] = `
ReduceResult {
  "data": Map {
    "香蕉" => ReduceResultItem {
      "value": ICUItem {
        "dumpStr": "{description, select, a{香蕉} b{香蕉} other{香蕉}}",
        "translates": Array [
          "香蕉",
        ],
      },
    },
    "苹果 {name}" => ReduceResultItem {
      "value": ICUItem {
        "dumpStr": "苹果 {name}",
        "translates": Array [
          "苹果 {name}",
        ],
      },
    },
    "西瓜" => ReduceResultItem {
      "value": ICUItem {
        "dumpStr": "西瓜",
        "translates": Array [
          "西瓜",
        ],
      },
    },
  },
}
`;

exports[`reduce 新增 & 删除 1`] = `
ReduceResult {
  "data": Map {
    "苹果 {name}" => ReduceResultItem {
      "value": ICUItem {
        "dumpStr": "苹果 {name}",
        "translates": Array [
          "苹果 {name}",
        ],
      },
    },
    "香蕉" => ReduceResultItem {
      "value": ICUItem {
        "dumpStr": "{description, select, a{香蕉} b{香蕉} c{香蕉}}",
        "translates": Array [
          "香蕉",
        ],
      },
    },
  },
}
`;
